pkg/safeguards/safeguards.go

```diff
@@ -0,0 +1,311 @@
+type ConstraintsBuilderA struct {
+}
+
+// fetchTestDeploymentFile pulls in our test deployment YAML
+func fetchDeploymentFile(path string) (map[string]interface{}, error) {
+	bs, err := os.ReadFile(path)
+	if err != nil {
+		// handle error
+		return nil, fmt.Errorf("reading deployment: %w", err)
+	}
+
+	var deploymentFile map[string]interface{}
+	if err := yaml.Unmarshal(bs, &deploymentFile); err != nil {
+		// handle error
+		return nil, fmt.Errorf("unmarshaling input: %w", err)
+	}
+
+	return deploymentFile, nil
+}
+
+type Params map[string]interface{}
+
+func buildParams() Params {
+	p := Params{
+		"allowedUsers": []string{
+			"nodeclient",
+			"system:serviceaccount:kube-system:aci-connector-linux",
+			"system:serviceaccount:kube-system:node-controller",
+			"acsService",
+			"aksService",
+			"system:serviceaccount:kube-system:cloud-node-manager",
+		},
+		"allowedGroups": []string{
+			"system:node",
+		},
+		"cpuLimit":           "200m",
+		"memoryLimit":        "1Gi",
+		"excludedContainers": []string{},
+		"excludedImages":     []string{},
+		"labels": []string{
+			"kubernetes.azure.com",
+		},
+		"allowedContainerImagesRegex": ".*",
+		"reservedTaints": []string{
+			"CriticalAddonsOnly",
+		},
+		"requiredProbes": []string{
+			"readinessProbe",
+			"livenessProbe",
+		},
+	}
+
+	return p
+}
+
+// thbarnes: placeholder for now
+func buildUserInfo() map[string]interface{} {
+	u := map[string]interface{}{
+		"username": "system:serviceaccount:kube-system:replicaset-controller",
+		"uid":      "439dea65-3e4e-4fa8-b5f8-8fdc4bc7cf53",
+		"groups": []string{
+			"system:serviceaccounts",
+			"system:serviceaccounts:kube-system",
+			"system:authenticated",
+		},
+	}
+
+	return u
+}
+
+// buildInput creates our input JSON when given a deployment file
+func buildInput(dfMap map[string]interface{}) map[string]interface{} {
+	// thbarnes: this needs to be refined and automated
+	// generated by kuberenetes api server -> not necessary for testing
+	// would be good to make the input a struct
+
+	input := map[string]interface{}{
+		"review": map[string]interface{}{
+			"object":   dfMap,
+			"userInfo": buildUserInfo(),
+		},
+		"parameters": buildParams(),
+	}
+	return input
+}
+
+func removeYAMLExtension(s string) string {
+	return strings.Split(s, ".yaml")[0]
+}
+
+// Option A's method of fetching constraints
+// thbarnes: figure out why consts aren't being read into debug instance
+// refine the path code
+func (cba ConstraintsBuilderA) Fetch() ([]ConstraintFile, map[string]map[string]interface{}, error) {
+	var c []ConstraintFile
+	var dfMap map[string]map[string]interface{}
+	cwd, _ := os.Getwd()
+	dirs := []string{cwd, ConstraintsDirectory}
+	fullPath := path.Join(dirs[0], dirs[1]) + "/"
+	constraintDirs, err := os.ReadDir(fullPath)
+	if err != nil {
+		return c, dfMap, fmt.Errorf("reading safeguards constraints directory for full list of dirs")
+	}
+
+	var constraints []os.DirEntry
+	for _, dir := range constraintDirs {
```

Comment:
you will save yourself a lot of pain in the future if you abstract this out into a go struct. This will also save you the trouble of walking a directory.

For example, consider having a `Safeguard` abstraction, it can look like this:
```golang
type Safeguard struct {
    name string
    filePath string //example: ./constraints/ContainerAllowedImages/container-allowed-images.yaml
}
```

This way you can define a list of safeguards such as:

```golang
var supportedSafeguards []Safeguard{}{
 {
     name: "ContainerAllowedImages",
     filePath: "path/to/yaml",
 },
// so on and so forth
}
```

Then here, you can simply loop through the list, and fetch the file. You no longer have to walk a directory path, nor account for the testcases folder

Reply:
done